mlir/lib/Bindings/Python/DialectLinalg.cpp

```diff
@@ -6,12 +6,45 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "mlir-c/BuiltinAttributes.h"
 #include "mlir-c/Dialect/Linalg.h"
 #include "mlir-c/IR.h"
-#include "mlir/Bindings/Python/NanobindAdaptors.h"
 #include "mlir/Bindings/Python/Nanobind.h"
+#include "mlir/Bindings/Python/NanobindAdaptors.h"
 
 namespace nb = nanobind;
+using namespace mlir::python::nanobind_adaptors;
+
+struct PyContractionDimensions {
+  MlirLinalgContractionDimensions value;
+
+  PyContractionDimensions() = default;
+  PyContractionDimensions(const MlirLinalgContractionDimensions &v)
+      : value(v) {}
+};
+
+static std::optional<PyContractionDimensions>
+mlirLinalgInferContractionDimensionsBinding(MlirOperation op) {
+  MlirLinalgContractionDimensions dims =
+      mlirLinalgInferContractionDimensions(op);
+
+  // Detect "empty" result.
+  if (mlirAttributeIsNull(dims.batch) && mlirAttributeIsNull(dims.m) &&
+      mlirAttributeIsNull(dims.n) && mlirAttributeIsNull(dims.k)) {
+    return std::nullopt;
+  }
+  return PyContractionDimensions{dims};
+}
+
+static std::vector<int32_t> convertDenseI32AttrToList(MlirAttribute attr) {
+  std::vector<int32_t> result;
+  int64_t size = mlirDenseArrayGetNumElements(attr);
+  result.reserve(size);
+  for (int64_t i = 0; i < size; ++i) {
+    result.push_back(mlirDenseI32ArrayGetElement(attr, i));
```

Comment:
```suggestion
  int64_t size = mlirDenseArrayGetNumElements(attr);
  std::vector<int32_t> result(size);
  for (int64_t i = 0; i < size; ++i) {
    result[i] = mlirDenseI32ArrayGetElement(attr, i);
```

Reply:
you actually don't need to do this at all - [PyDenseI32ArrayAttribute](https://github.com/makslevental/llvm-project/blob/5d3ae5161210c068d01ffba36c8e0761e9971179/mlir/lib/Bindings/Python/IRAttributes.cpp#L432) is already mapped so if you just return the "bare" `batch`, `m`, etc below they should pop out in Python as objects that supports [element-wise access already](https://github.com/makslevental/llvm-project/blob/5d3ae5161210c068d01ffba36c8e0761e9971179/mlir/lib/Bindings/Python/IRAttributes.cpp#L366).